openmdao/visualization/meta_model_viewer/meta_model_visualization.py

```diff
@@ -0,0 +1,551 @@
+
+
+    # Event handler functions
+    def update(self, attr, old, new):
+        self.update_all_plots()
+
+    def scatter_plots_update(self, attr, old, new):
+        self.update_subplots()
+
+    def scatter_input(self, attr, old, new):
+        self.dist_range = float(new)
+        self.update_all_plots()
+
+    def input_dropdown_checks(self,x,y):
+        # Checks to see if x and y inputs are equal to each other
+        if x == y:
+            return False
+        else:
+            return True
+    
+    def x_input_update(self, attr, old, new):
+        if self.input_dropdown_checks(new, self.y_input.value) == False:
+            raise ValueError("Inputs should not equal each other")
+        else:
+            self.update_all_plots()
+
+    def y_input_update(self, attr, old, new):
+        if self.input_dropdown_checks(self.x_input.value, new) == False:
+            raise ValueError("Inputs should not equal each other")
+        else: 
+            self.update_all_plots()
+
+    def output_value_update(self, attr, old, new):
+        self.output_variable = self.output_list.index(new)
+        self.update_all_plots()
+
+    def training_points(self):
+
+        """
+        Calculates the training points and returns and array containing the position and alpha opacity of the training points nearest to the surrogate line.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        array
+            The array of training points and their alpha opacity with respect to the surrogate line
+        """
+
+        # xt contains
+        # [x1, x2, x3, x4]
+        xt = self.surrogate_comp._training_input # Input Data
+        yt = np.squeeze(self.stack_outputs(self.surrogate_comp._training_output), axis=1) # Output Data
+        output_variable = self.output_list.index(self.output_select.value)
+        data = np.zeros((0, 8))
+
+        # Calculate the limits of each input parameter
+        bounds = [[min(i), max(i)] for i in self.input_data.values()]
+        limits = np.array(bounds)
+        self.limit_range = limits[:, 1] - limits[:, 0]
+
+        # Vertically stack the x/y inputs and then transpose them 
+        infos = np.vstack((xt[:, self.x_index], xt[:, self.y_index])).transpose()
+        points = xt.copy()
+        # Set the first two columns of the points array to x/y inputs, respectively
+        points[:, self.x_index] = self.input_point_list[self.x_index]
+        points[:, self.y_index] = self.input_point_list[self.y_index]
+        points = np.divide(points, self.limit_range)
+        tree = cKDTree(points)
+        dist_limit = np.linalg.norm(self.dist_range * self.limit_range)
+        scaled_x0 = np.divide(self.input_point_list, self.limit_range)
+        # Query the nearest neighbors tree for the closest points to the scaled x0 array
+        dists, idx = tree.query(scaled_x0, k=len(xt), distance_upper_bound=dist_limit)
+        idx = idx[idx != len(xt)]
```

Comment:
this for loop looks wonky to me. Why do you assign idx like this on line 514, then use an if condition in the for loop right after.